_posts/2016-03-01-Docker-meets-Continuous-Deployment.md

```diff
@@ -0,0 +1,165 @@
+---
+layout: post
+name: docker-meets-continuous-deployment
+title: Docker meets Continuous Deployment
+date: 2016-03-01 00:00:00 +00:00
+author: Robert Firek
+image:
+    src: /assets/img/custom/blog/2016-03-01-docker-meets-continuous-deployment.png
+    attribution:
+        text: Track icon by Freepik (CC 3.0 BY)
+        href: http://www.freepik.com
+tags:
+- Continuous Deployment
+- Docker
+- TeamCity
+--- 
+
+About one year ago I had a first contact with Docker. This new kid on the block promised to relieve our poor computers from installation of all tools, languages, dependencies and operating systems. Isolated run environments emerged on developers computers.
+
+A conservative approach of my ops teammates didn't change my joy of using Docker. Despite many people describing Docker as tool writen by developers for development our industry found new ways of using images and containers. Images of our applications and services became units of deployment for tools like Kubernetes, Docker Swarm or Marathon.
+
+But how this images appear in these tools?
+
+## Setting The Scene
+
+From developer's perspective any application is manifested by a code. There is long way before our code find a good place to execute on a production environment. I want to show how you this process can be easier with **Docker** and **Continuous Deployment** pipeline. To do that we need a small application. Our application is going to expose some HTTP API which we are going to call after the deployment. Let's assume that will are going to use **Gradle** to build the application and **TeamCity** for build automation.
+
+On each **TeamCity** build agent we should have installed **Docker**. We will also use this machine to run our application. In a real project we don't want to install **TeamCity** agents on all machines and we should introduce another tool (e.g. Kubernetes) which will take care about distribution of our application.
+ 
+## Build
+
+We need to build something before we will be able to create any docker image. To do that we need to create the first step in our Continuous Deployment (CD) pipeline. 
+In this step we will download our source code, run all **tests** which can be done on the source code level and produce an **artifact** containing all elements required to start and run our application.
+
+This build configuration is not very different to a usual step in a common Continuous Deployment pipeline. Alongside common parameters we have to define **artifacts** which will be generated after each execution. We are going to use them as a base for next steps in the pipeline.
```

Comment:
@sandromancuso I want to say that you always have to build something from your code, before you release anything, so this is step is common for all CD pipelines. Next we have to define common parameters like a name, but in this case we have to define also artifacts which we want to produce in this step.